### PR TITLE
new JUnit5SuiteMisuse error-prone check

### PR DESCRIPTION
**Pull request by iamdanfox**
## Before this PR

I started building some defensiveness in another PR (https://github.com/palantir/gradle-baseline/pull/837) but it turns out it didn't actually catch the problem that @delta003 noticed.

His case was that he still had a floating `@RunWith(Suite.class)` class (Junit4) but it _referenced_ classes containing JUnit5 tests, leading to tests _silently not running_ 🔥🔥🔥

## After this PR
==COMMIT_MSG==
ErrorProne will now detect dangerous usage of `@RunWith(Suite.class)` that references JUnit5 classes, as this can cause tests to silently not run!
==COMMIT_MSG==

## Possible downsides?
- I wonder if accumulating things in a hashmap will end up OOMing??

### CODE REVIEW

File: baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/JUnit5SuiteMisuse.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+        severity = BugPattern.SeverityLevel.ERROR,
+        summary = "Referencing JUnit5 tests from JUnit4 Suites will silently not work")
+public final class JUnit5SuiteMisuse extends BugChecker
+        implements BugChecker.ClassTreeMatcher, BugChecker.AnnotationTreeMatcher {
+
+    private static final long serialVersionUID = 1L;
+
+    // We remember classes and validate them later because error-prone doesn't let us arbitrarily explore classes we
+    // discover when reading the @SuiteClasses annotation.
+    private static final Set<Type.ClassType> knownJUnit5TestClasses = new HashSet<>();
```
Comment (ferozco) [inline baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/JUnit5SuiteMisuse.java:52]: This seems pretty weird, but I can't think of another alternative.

File: baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/JUnit5SuiteMisuse.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+                return buildDescription(tree)
+                        .setMessage("Don't reference JUnit5 test classes from JUnit4 SuiteClasses annotation")
+                        .build();
+            }
+        }
+
+        return Description.NO_MATCH;
+    }
+
+    private static List<JCTree.JCFieldAccess> getReferencedClasses(AnnotationTree tree, VisitorState state) {
```
Comment (ferozco) [inline baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/JUnit5SuiteMisuse.java]: `List<ClassType>` instead?

File: baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/StringBuilderConstantParameters.java [hunk 3]
```diff
@@ -45,6 +45,8 @@
 @AutoService(BugChecker.class)
 @BugPattern(
         name = "StringBuilderConstantParameters",
+        link = "https://github.com/palantir/gradle-baseline#baseline-error-prone-checks",
```
Comment (ferozco) [inline baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/StringBuilderConstantParameters.java:48]: unrelated but nice

**Review by ferozco — COMMENTED** [review]
Could we test this out against some repos just to double check everythings fine

**Comment by ferozco** [comment]
👍

**Comment by iamdanfox** [comment]
Verified on scheduler, this would catch the specific problem that @delta003 had:

```
./gradlew scheduler:integrationTestClasses

> Task :scheduler:compileIntegrationTestJava
/Volumes/git/scheduler/scheduler/src/integrationTest/java/com/palantir/scheduler/resource/SchedulerResourceIntegrationTest.java:46: error: [JUnit5SuiteMisuse] Class uses JUnit5 tests but is referenced by a JUnit4 SuiteClasses
public final class SchedulerResourceIntegrationTest extends SchedulerBaseIntegrationTest {
             ^
    (see https://github.com/palantir/gradle-baseline#baseline-error-prone-checks)
1 error

> Task :scheduler:compileIntegrationTestJava FAILED
```

File: baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/JUnit5SuiteMisuse.java (hunk 1 above)
Comment (iamdanfox) [reply baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/JUnit5SuiteMisuse.java:52]: Agreed - I tried a bunch of things before landing on this. You can traverse up your class hierarchy nice and happily, and you can do whatever you like inside a file, I think they enforce that you can't go spelunking around arbitrary classes to ensure things are performant.

**Pull request merged**